lib/Sema/TypeCheckDecl.cpp

```diff
@@ -1190,71 +1191,54 @@ class TypeAccessScopeChecker : private TypeWalker {
   using TypeAccessScopeCacheMap = TypeChecker::TypeAccessScopeCacheMap;
   TypeAccessScopeCacheMap &Cache;
   const SourceFile *File;
-  SmallVector<intptr_t, 8> RawScopeStack;
-
-  static constexpr const intptr_t INVALID = -1;
+  SmallVector<Optional<AccessScope>, 8> RawScopeStack;
 
   explicit TypeAccessScopeChecker(TypeAccessScopeCacheMap &cache,
                                   const SourceFile *file)
       : Cache(cache), File(file) {
     // Always have something on the stack.
-    RawScopeStack.push_back((intptr_t)INVALID);
+    RawScopeStack.push_back(None);
   }
 
   bool shouldVisitOriginalSubstitutedType() override { return true; }
 
-  static intptr_t intersectAccess(const intptr_t first, const intptr_t second) {
-    if (first == INVALID || second == INVALID)
-      return INVALID;
-    if (!first)
-      return second;
-    if (!second)
-      return first;
-    if (first == second)
-      return first;
-
-    auto firstDC = reinterpret_cast<const DeclContext *>(first);
-    auto secondDC = reinterpret_cast<const DeclContext *>(second);
-    if (firstDC->isChildContextOf(secondDC))
-      return first;
-    if (secondDC->isChildContextOf(firstDC))
-      return second;
-    return INVALID;
-  }
-
   Action walkToTypePre(Type ty) override {
     // Assume failure until we post-visit this node.
     // This will be correct as long as we don't ever have self-referential
     // Types.
     auto cached = Cache.find(ty);
     if (cached != Cache.end()) {
-      auto opaqueCached = reinterpret_cast<intptr_t>(cached->second);
-      RawScopeStack.back() = intersectAccess(RawScopeStack.back(),opaqueCached);
+      auto last = &RawScopeStack.back();
```

Comment:
Nitpick: I think we'd prefer `auto &last = RawScopeStack.back()` or even `AccessScope &last = RawScopeStack.back()`, because it won't be reassigned. (Same below.)